diff/diff.go

```diff
@@ -163,11 +164,29 @@ func redactSecrets(old, new *manifest.MappingResult) {
 		if err := yaml.NewYAMLToJSONDecoder(bytes.NewBufferString(old.Content)).Decode(&oldSecret); err != nil {
 			old.Content = fmt.Sprintf("Error parsing old secret: %s", err)
 		}
+		if oldSecret.Data == nil {
+			oldSecret.Data = make(map[string][]byte, len(oldSecret.StringData))
+		}
+		for k, v := range oldSecret.StringData {
+			data := []byte(v)
+			dst := make([]byte, base64.StdEncoding.EncodedLen(len(data)))
+			base64.StdEncoding.Encode(dst, data)
+			oldSecret.Data[k] = dst
+		}
 	}
 	if new != nil {
 		if err := yaml.NewYAMLToJSONDecoder(bytes.NewBufferString(new.Content)).Decode(&newSecret); err != nil {
 			new.Content = fmt.Sprintf("Error parsing new secret: %s", err)
 		}
+		if newSecret.Data == nil {
+			newSecret.Data = make(map[string][]byte, len(newSecret.StringData))
+		}
+		for k, v := range newSecret.StringData {
```

Comment:
I might be confused, but shouldn't we reset the whole `newSecret.Data` before merging in StringData entries? Otherwise, it might get redundant keys when you have different set of keys for data and stringData respectively?

In other words, what if it looked like:

```
data:
  foo: <BASE64 encoded value "FOO">
  bar: <BASE64 encoded value "BAR1">
stringData:
  bar: BAR2
  baz: BAZ
```

Should this function output:

```
stringData:
  foo: FOO
  bar: BAR2
  baz: BAZ
```

or

```
stringData:
  bar: BAR2
  baz: BAZ
```

?

Reply:
This is to handle the case where the incoming YAML manifest looks as follows:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: example
data:
  foo: <encoded:FOO>
  bar: <encoded:BAR>
stringData:
  bar: BAR2
  baz: BAZ
```

In such a case, per the Kubernetes documentation, the effective value of such a manifest is as follows:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: example
data:
  foo: <encoded:FOO>
  bar: <encoded:BAR2>
  baz: <encoded:BAZ>
```
... or, equivalently:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: example
stringData:
  foo: FOO
  bar: BAR2
  baz: BAZ
```

Reply:
@mumoshu Specifically, reference the following from the [Kubernetes documentation](https://kubernetes.io/docs/concepts/configuration/secret/#restriction-names-data):
> All key-value pairs in the stringData field are internally merged into the data field. If a key appears in both the data and the stringData field, the value specified in the stringData field takes precedence.

Reply:
@mumoshu Were there any other questions you had about this?